pkg/commands/docker.go

```diff
@@ -67,9 +73,134 @@ func (c *DockerCommand) NewCommandObject(obj CommandObject) CommandObject {
+func handleSSHDockerHost() error {
+	const key = "DOCKER_HOST"
+	ctx := context.Background()
+	u, err := url.Parse(os.Getenv(key))
+	if err != nil {
+		// if no or an invalid docker host is specified, continue nominally
+		return nil
+	}
+
+	// if the docker host scheme is "ssh", forward the docker socket before creating the client
+	if u.Scheme == "ssh" {
+		newDockerHost, err := tunneledDockerHost(ctx, u.Host)
+		if err != nil {
+			return fmt.Errorf("tunnel ssh docker host: %w", err)
+		}
+		err = os.Setenv(key, newDockerHost)
+		if err != nil {
+			return fmt.Errorf("override DOCKER_HOST to tunneled socket: %w", err)
+		}
+
+		return nil
+	}
+	return nil
+}
+
+func tunneledDockerHost(ctx context.Context, remoteHost string) (string, error) {
+	socketDir, err := ioutil.TempDir("/tmp", "lazydocker-sshtunnel-")
+	if err != nil {
+		return "", fmt.Errorf("create ssh tunnel tmp file: %w", err)
+	}
+	localSocket := path.Join(socketDir, "dockerhost.sock")
+
+	err = tunnelSSH(ctx, remoteHost, localSocket)
+	if err != nil {
+		return "", fmt.Errorf("tunnel docker host over ssh: %w", err)
+	}
+
+	// set a reasonable timeout, then wait for the socket to dial successfully
+	// before attempting to create a new docker client
+	const socketTunnelTimeout = 8 * time.Second
+	ctx, cancel := context.WithTimeout(ctx, socketTunnelTimeout)
+	defer cancel()
+
+	err = retrySocketDial(ctx, localSocket)
+	if err != nil {
+		return "", fmt.Errorf("ssh tunneled socket never became available: %w", err)
+	}
+
+	// construct the new DOCKER_HOST url with the proper scheme
+	newDockerHostURL := url.URL{Scheme: "unix", Path: localSocket}
+	return newDockerHostURL.String(), nil
+}
+
+// Attempt to dial the socket until it becomes available.
+// The retry loop will continue until the parent context is canceled.
+func retrySocketDial(ctx context.Context, socketPath string) error {
+	t := time.NewTicker(1 * time.Second)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-t.C:
+		}
+		// attempt to dial the socket, exit on success
+		err := tryDial(ctx, socketPath)
+		if err != nil {
+			continue
+		}
+		return nil
+	}
+}
+
+// Try to dial the specified unix socket, immediately close the connection if successfully created.
+func tryDial(ctx context.Context, socketPath string) error {
+	var dialer net.Dialer
+	conn, err := dialer.DialContext(ctx, "unix", socketPath)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	return nil
+}
+
+// CloseDockerSocketConnection kills the docker socket SSH forwarding process, if it exists.
+//
+// If will exist when DOCKER_HOST has the protocol scheme `ssh://`.
+func CloseDockerSocketConnection() {
+	if dockerSSHConnection != nil {
+		syscall.Kill(-dockerSSHConnection.Process.Pid, syscall.SIGKILL)
+	}
+}
+
+// dockerSSHConnection holds package-level state for the last-opened SSH tunnel to a remote docker socket.
+var dockerSSHConnection *exec.Cmd
```

Comment:
What are your thoughts on having this be a field on a struct that we modify, as opposed to a package-level variable? Reason being is that in my experience all package level variables inevitably cause problems for me. We could potentially store this on our App struct, a DockerCommand struct, or some new struct that contains all of these newly added methods.

Reply:
I completely agree that package-level state is not a good solution here. I just pushed a change that abstracts the docker socket cleanup under the chain of `Close` methods... your call whether this is in fact cleaner https://github.com/jesseduffield/lazydocker/pull/268/commits/afe3d23cbd0345dfe786f4efa29283338dff5e79